File: Android/app/src/main/java/com/dreamproject/inwords/presentation/viewScenario/octoGame/BaseSingleTypeAdapter.java
package com.dreamproject.inwords.presentation.viewScenario.octoGame;

import android.util.Pair;
import android.view.LayoutInflater;

import java.util.ArrayList;
import java.util.List;

import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;

public abstract class BaseSingleTypeAdapter<T, VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH>
        implements Consumer<Pair<List<T>, DiffUtil.DiffResult>> {
    protected final LayoutInflater inflater;
    protected List<T> values;
    protected PublishSubject<T> onItemClickedListener;

    protected BaseSingleTypeAdapter(List<T> values,
                                    LayoutInflater layoutInflater,
                                    PublishSubject<T> onItemClickedListener) {
        this.inflater = layoutInflater;
        this.values = values;
        this.onItemClickedListener = onItemClickedListener;
    }

    protected BaseSingleTypeAdapter(LayoutInflater layoutInflater,
                                    PublishSubject<T> onItemClickedListener) {
        this(new ArrayList<>(), layoutInflater, onItemClickedListener);
    }

    @Override
    public void accept(Pair<List<T>, DiffUtil.DiffResult> pair) {
        this.values = pair.first;
        pair.second.dispatchUpdatesTo(this);
    }

    public List<T> getValues() {
        return values;
    }

    @Override
    public int getItemCount() {
        return values.size();
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/presentation/viewScenario/octoGame/BaseSingleTypeAdapter.kt
package ru.inwords.inwords.presentation.viewScenario.octoGame

import android.view.LayoutInflater
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.functions.Consumer
import io.reactivex.subjects.Subject

abstract class BaseSingleTypeAdapter<T : Any, VH : RecyclerView.ViewHolder>
protected constructor(protected val layoutInflater: LayoutInflater,
                      protected val onItemClickListener: Subject<T>)
    : RecyclerView.Adapter<VH>(), Consumer<Pair<List<T>, DiffUtil.DiffResult>> {

    var items: List<T> = emptyList()
        protected set

    override fun accept(pair: Pair<List<T>, DiffUtil.DiffResult>) {
        this.items = pair.first
        pair.second.dispatchUpdatesTo(this)
    }

    override fun getItemCount(): Int {
        return items.size
    }
}
